### DISCUSSION TITLE
How to replicate a recoil atom family?

### DISCUSSION THREAD

**Discussion by Cpt-Falcon** [pmndrs/jotai#1850] [Q&A]
Hello all,

I'm trying to follow the example on Jotai for the atom family utility:

import { atomFamily } from 'jotai/utils';

I'm coming from recoil so I don't really understand how to replicate the behavior or what I should be doing. For example, here is what I tried to do:
```
export const stateSystemParameters = atomFamily(
    ({ guid, system  }: { guid: string; system: System}) => atom<System>(system),
    (x, y) => x.guid === y.guid,
);

```
In this case, I'm trying to map a guid to a typescript class called system. I want to pass in a guid and get the atom label back so that I can manipulate it where I need to. But I'm always forced to pass in the guid and the label, even if I don't set the label. I think I'm misunderstanding something fundamental here. I want the atom to contain system, but I want the equality family check to work based on the guide id. I tried this as well:

```
export const stateSystemParameters = atomFamily(
    (guid: string) => atom(guid),
    (x, y) => x === y,
);

```
But now it doesn't have the right type and it's being configured for string type, not system. I would appreciate any input or advice here on how to deal with this, or if I'm going about this the wrong way. Thank you for your support.

Here's the example of the corresponding recoil code I'm trying to port:

```
export const stateSystemParameters = atomFamily<System, string>({
    key: 'state-system-parameters',
});
```

Edit:

This feels quite hacky, but I guess it would work:

```
export const stateSystemParameters = atomFamily(
    ({ guid, system = null }: { guid: string; system: System | null}) => atom<System>(system ?? systemDefaults.clone()),
    (x, y) => x.guid === y.guid,
);
```

**Comment by dai-shi** [comment]
Thanks opening up a discussion.

A general guide for anyone is, although "atomFamily" name is originally taken from Recoil, its implementation and the concept is different.

In Jotai, atomFamily is a `Map`. It's probably easier to learn it from the simplified code.
https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```

**Reply by Cpt-Falcon** [reply]
Thank you for your response, this is very helpful.

**Comment by dai-shi** [comment]
> This feels quite hacky, but I guess it would work:
> 
> ```
> export const stateSystemParameters = atomFamily(
>     ({ guid, system = null }: { guid: string; system: System | null}) => atom<System>(system ?? systemDefaults.clone()),
>     (x, y) => x.guid === y.guid,
> );
> ```

That looks pretty good.

**Reply by Cpt-Falcon** [reply]
Awesome, I was worried I was using the atomFamily in a way that wasn't intended. Thank you for the feedback.
